address #1981 with SchXslt (aka mausatron) instead of skeleton

sydb opened the pull request:
I will say more shortly, but the main point now is that this PR updates our build process to use @dmj’s `SchXslt` (which I call “mausatron”) rather than the original skeleton implementation. This has quite a few minor advantages, but mainly allows users (including us) to
1. use abstract patterns; and
2. use an attribute as a context node. 

Reviewers — the lastest merge from dev into this branch (issue1981bis) required that the various Test/expected_results/detest*log files be hand-merged. Thus it is appropriate to take a pretty careful look at these.

sydb commented:
I just merged the current dev branch into this branch, _Guidelines_ still build (include `make test`) without error.
I am kinda hoping this could get pulled in before our Council meeting on Fri 09 Feb 24. Pipe dream or realistic?

joeytakeda commented:
There seems to be an error with (I believe) the NVDL part of the build (note that this is not causing the build to fail, but is producing error messages—you can see it in the Actions build as well as when run locally):

```
BUILD: Check validity with nvdl, first examples with feasible validity, and then the valid ones
./run-onvdl p5.nvdl p5.xml
Schema implementation error: This schema has abstract patterns, yet they are supposed to be preprocessed out already
             
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
./run-onvdl p5valid.nvdl v.xml
Schema implementation error: This schema has abstract patterns, yet they are supposed to be preprocessed out already
             
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam
Error: unrecognized element in ISO Schematron namespace: check spelling
			and capitalizationparam

```

sabineseifert commented:
Full Council discussion on April 13 at VF2F:

European subgroup was getting the same errors.
Several options discussed:
1. find someone to fix nvdl validation
- If we continue to use NVDL, then we will need to fix the production of the NVDL in the release, e.g. https://tei-c.org/Vault/P5/4.1.0/xml/tei/Exemplars/tei_all.nvdl does not work as it cannot find referenced resources (e.g. ../schematron.rng)
2. Stop using nvdl entirely:
- remove the nvdl files (because they just don't work)
- Or stop using the nvdl files in build process
3. Continue using nvdl with line 105 commented out, but then need to fix the produced NVDL documents

**Decision**: find someone to fix nvdl validation
Get update on oNVDL at Balisage. If there has been no further work on oNVDL then we could proceed with getting rid of it since no-one is using it.

sydb commented:
> Get update on oNVDL at Balisage. If there has been no further work on oNVDL then we could proceed with getting rid of it since no-one is using it.

I was unsuccessful in finding out about oNVDL support at _Balisage_. Next step is to ask on XML Slack, or no repo, or direct to author(s).

sydb commented:
Whoo-hoo! :man_dancing: **major** progress on this ticket. First the routine stuff:

1. Merged dev branch into this one, resolving a few conflicts by hand and re-producing some expected results.¹
2. Updated Mausatron from 1.9.5 to 1.10.1.

Now the exciting stuff. I have figured out the problem, and a large part of the solution, to the NVDL problem we were having. The lesson is to **read the $*@%ing error msg**! Well over a year ago @joeytakeda pasted in [the errors](https://github.com/TEIC/TEI/pull/2509#issuecomment-1936205523), but we all looked at the long list, not the explanatory message that precedes it: “This schema has abstract patterns, yet they are supposed to be preprocessed out already”. 

NVDL is telling us what we have to do — pre-process the Schematron file so that abstract patterns are processed into concrete patterns before NVDL is run. That means there is probably a (relatively easy) way to do that. I poked around in the Mausatron code, and found that expand.xsl seems to do exactly the right thing. I added a call to it to antbuilder.xml (so now there are 5 steps to create p5odds.isosch.xsl, numbered 10a–10e), and _most_ the build process seems to work!

That’s the good news. The bad news is:

1. I am not at all sure I am calling expand.xsl the right way — I think maybe I was just supposed to use pipeline.xsl. But that is pretty easy to fix.
2. The build process now dies on simplePrint, because it deletes the att.declarable class, which is the one that contains the abstract pattern. But every element that is a member of that class has an instance of that pattern. I am not sure what the right solution to this is, yet. (And I don’t think there are any easy solutions.)

**Note**
¹ And forgot that if there is a difference between detest_odd_schematron.txt and expected_results/detest_odd_schematron.txt, and you determine the current version is correct and should appear in expected_results/, the file you need to copy over is detest_odd_schematron.**svrl**, not .txt.

sydb commented:
So I have thought through the problems a little more:

1. Yes, I am pretty sure I am using expand.xsl the right way.
2. Since Schematron _requires_ that if there is a `<pattern is-a="duck">` there must also be a `<pattern abstract="true" id="duck">`, the solution (I think) will be to have extract_isosch.xsl only extract those is-a patterns that have a matching abstract pattern.

sydb commented:
### next steps

We need two new issues:

1. Updating extract-isosch.xsl so that it does _not_ extract a pattern with `@is-a` unless it has also extracted a corresponding abstract pattern. I plan to open the ticket, do the work, and submit PR for this myself. (I have not decided if that should be a PR against dev or against this branch — any thoughts?) — see TEIC/Stylesheets#749
2. Updating the regular Stylesheets (teiodds.xsl, I think) to do the same thing when generating Schematron-inside-RELAX. I plan to open the ticket myself, but am hoping someone else steps up to do the work, as I am not very familiar with that code. — see TEIC/Stylesheets#750

martindholmes reviewed (approved):
This looks good to me; tests pass OK. I'm running a full build now as an additional test.
